connlimit.go

```diff
@@ -182,11 +192,29 @@ func (l *Limiter) SetConfig(c Config) {
 // their own calls if they need to continue limiting the number of concurrent
 // hijacked connections.
 func (l *Limiter) HTTPConnStateFunc() func(net.Conn, http.ConnState) {
+	message := "Your IP is issuing too many concurrent connections, please rate limit your calls\n"
+	tooManyRequestsResponse := []byte(fmt.Sprintf("HTTP/1.1 429 Too Many Requests\r\n"+
+		"Content-Type: text/plain\r\n"+
+		"Content-Length: %d\r\n"+
+		"Connection: close\r\n\r\n%s", len(message), message))
+	// Buffer shared by all clients, we don't care about result
+	readBuffer := make([]byte, 0, 4096)
```

Comment:
The `conn.Read(readBuffer)` below will only read up to the *length* of this slice, so 0. From my testing this is no different than not performing a read-before-write at all.

Switching this to `make([]byte, 4096)` seems to properly read incoming headers before writing the response. However, sharing that buffer seems like a data-race (although I have been unable to get the race detector to complain).

I made a minimal repro case to play around: https://raw.githubusercontent.com/schmichael/play/53174a690f670b8cca6e5275530baf9119a9834c/nethttp/nethttp.go

Run with `go run -race nethttp.go`

Flip these 2 lines around to test the various buffering approaches: https://github.com/schmichael/play/blob/53174a690f670b8cca6e5275530baf9119a9834c/nethttp/nethttp.go#L74-L75

Reply:
You are right, was not sure about the need for Read() ... the best is probably to remove it

Reply:
Just removed it, does not seems to change the behavior with curl, this was probably just overengineering.